### ISSUE TITLE
Create IDs for speeches

### ISSUE THREAD

**Issue by ninpnin** [swerik-project/riksdagen-records#13]
Current options
- List speeches in the metadata block
- Wrap speeches in div elements

**Comment by BobBorges** [comment]
I think wrapping speeches in divs is short sighted for a "living" resource -- we don't know how many things like this will be tagged and whether all potential things to tag will allow the hierarchical structure required by xml divs. Metadata blocks would allow tagging whatever features independently of other tagged features without creating a bottomless pit of divs.

**Comment by ninpnin** [comment]
I just realized we could use the n attributes that are available for all elements. From the documentation,

>n (number) gives a number (or other label) for an element, which is not necessarily unique within the document. 

Then, we would just include the ID in all u elements that belong to the speech. Eg. for the following speech with the ID  i-AzXa4EUmTu6mz8YQsCpizb

```xml
<note type="speaker" xml:id="i-G36fJpDJFVqwFFQjbknRq2">
  Herr ERIKSSON i Bäeckmora (cp):
</note>
<u xml:id="i-3KxGSd288AdTa9bfy9BtMv" xml:n="i-AzXa4EUmTu6mz8YQsCpizb" next="i-QzTu4nNrn4q8kU1N1u4xZC" who="i-7CXHDen9y2qKcYDisT3zjQ">
  <seg xml:id="i-EV3wMeu3xQ8QuzNwmvWbjM">
    Herr talman! I det som statsrådet Palme sade nu fanns väl egentligen
    [...]
    som en stor del av svenska folket bestämt önskar få ändring i.
  </seg>
</u>
<u xml:id="i-QzTu4nNrn4q8kU1N1u4xZC" xml:n="i-AzXa4EUmTu6mz8YQsCpizb" prev="i-3KxGSd288AdTa9bfy9BtMv" who="i-7CXHDen9y2qKcYDisT3zjQ" next="i-CUrwEDJ9XoTNrw9wfqWrYb">
  <seg xml:id="i-4mZ33Z1km8JtDLieMQPm5Q">
    Jag anser att statsrådet Palme på denna punkt också skulle uppta
    allvar-
  </seg>
</u>
```

**Comment by BobBorges** [comment]
What happens when this same fragment gets tagged as multiple things? Do we have multiple n attribs, or multiple IDs in the n?

**Comment by ninpnin** [comment]
Which fragment are you referring to?

**Comment by MansMeg** [comment]
We should follo the TEI guidelines. n should be used for page number.
https://www.tei-c.org/release/doc/tei-p5-doc/en/html/TS.html#TSBAUT
https://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-u.html

**Comment by ninpnin** [comment]
Where does it say that? [Here](https://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-att.global.html) it says

>(number) gives a number (or other label) for an element, which is not necessarily unique within the document. 

It might be a page number for pb elements, but for u elements I find no such description.

**Comment by BobBorges** [comment]
@ninpnin I refer to the fragment you posted as an example. It's tagged as a speech with ID, but down the line it may be tagged with other things... an interpellation debate, or some other type of sectioning that may or may not coincide exactly with the speech itself. So how does the approach you describe handle multiple possible xml:n values?

**Comment by ninpnin** [comment]
@BobBorges Debates are more suited for div-wrapping, along with any non-overlapping sectioning. But if we have other possibly overlapping things, I unfortunately have no solution for that.

**Comment by BobBorges** [comment]
It seems like putting these thing as element lists in the tei header would be most flexible, and cleanest in the case when a human has to look at the xml.

**Comment by ninpnin** [comment]
does the schema allow for that?

**Comment by MansMeg** [comment]
I agree with Bob, that for now the solution "List speeches in the metadata block" sounds like the best one. Would that work with the TEI schema?

I also added a third option. that is to make each speech only be one block and then rather have paragraph breaks within each utterance. It is semantically closer to the TEI schema than how we solve it now (and the id of the u block would be the speech ID). But still, I think the first solution is best.

**Comment by BobBorges** [comment]
there are a couple of options (parlaclarin):
• `<TEI><standOff>` contains all kinds of stuff that could be useful here
• `<teiHeader><profileDesc><textDesc>` has domain, interaction, purpose
• `<teiHeader>` has a <xenoData> elem which takes any kind of metadata in whatever format (dangerously flexible :) )

From the parlaclarin given examples, looks like standOff is the closest to what we want, but we could also consider like listGrp with type attrib, id, and sub elems that contain a referring ID for the segs we want to label.

**Comment by MansMeg** [comment]
- [ ] Open up an issue on how to store speech ids in ParlaClarin 
- [ ] Write down a decision on how to do it
- [ ] Add it to the corpus

**Comment by ninpnin** [comment]
Follow at https://github.com/clarin-eric/parla-clarin/issues/25

**Comment by BobBorges** [comment]
It seems like the parlaclarin people aren't eager to get involved in this. I propose a workaround (or workwithin?) here:

Under `teiHeader/profileDesc/textDesc/constitution`, which "describes the internal composition of a text" we can add  `note` elements that contain `desc` (description) `linkGrp` (link group) and `ptr` (pointer) elements. I'm attaching a minimal working example here that passes parlaclarin and tei validation. 

![Image](https://github.com/user-attachments/assets/6dc96854-8db5-47d5-8aad-9301e5d0f236)


This strategy would allow several nice improvements (1) speech Ids (in the example, it's the id attribute of notes with type "speech") (2) labeling and categorizing the text without cluttering up the `body` descendants or getting out of hand with recursion depth (3) labeling and classifying non-hierarchical and discontinuous features of the text (e.g. a debate that was stopped to attend to some other business and taken up later) (4) easier search and extract of particular types of text (If I'm only interested in speeches from interpellation debates, I parse the header, get `u` elem IDs from the IP speeches and .find() them instead of iterating over the entire protocol).

[_tei_heder_speech_id.xml.txt](https://github.com/user-attachments/files/17341551/_tei_heder_speech_id.xml.txt)

**Comment by fredrik1984** [comment]
Ok, I think it is excellent that we get IDs for speeches. @MansMeg and @ninpnin, do you have any thoughts on the technical aspects here?

**Comment by ninpnin** [comment]
Looks good. I would maybe only add the paragraph IDs, so that it doesn't get too cluttered, but that's a minor consideration and a matter of taste.

@BobBorges Does the example adhere to TEI and ParlaClarin? If not the latter, we could suggest that to the parlaclarin people.

**Comment by BobBorges** [comment]
It passes verification -- both TEI and parlaclarin. So if we're happy with it, I think we don't need to do anything else... or maybe show them the example in case some other people want a speech ID.

**Comment by ninpnin** [comment]
Now we have the notation. Another consideration is, what do we define as a speech?

Do we regenerate indices every time a paragraph is added or deleted? Or perhaps every time the _first_ or the _last paragraph is changed?

**Comment by MansMeg** [comment]
Oh. Im not sure I understand fully how this will work. It seems from you example that you also cluster the segments into debates? I think it is better to keep this clear and only have a speech id and then ptr to all the segments that belong to the same speech?

I think the best approach is to make this conclusion as a decision with an example, etc. I.e. file a PR for this to the riksdagen-records repo?

**Comment by fredrik1984** [comment]
ok, that sounds good! the important thing for now is to be able to identify and annotate individual speeches. but it is good that the structure also allows for annotating speeches to specific debates.

**Comment by BobBorges** [comment]
yes! I think the goal down the line is to ID all "types" of debates and addresses within the protocols as we have started doing with interpellation debates. If we implement a strategy like this, all the action happens in the metadata and we don't have to worry about messing up the text, how to limit recursion in the xml, or any similar problems.

**Comment by BobBorges** [comment]
... I will move this to a decision proposal sometime today

**Comment by MansMeg** [comment]
Now @BobBorges will implement :
- [ ] the decision as IDs in the corpus as a new PR.
- [ ] add unit test that speeches has utterances and all records has speeches and there is equally many introductions as speeches
